src/main/java/io/cdap/plugin/http/source/common/pagination/page/JsonPage.java

```
@@ -0,0 +1,83 @@
+import com.google.gson.JsonObject;
+import io.cdap.plugin.http.source.common.BaseHttpSourceConfig;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Returns elements from json one by one by given json path.
+ */
+class JsonPage implements BasePage {
+  private final String insideElementJsonPathPart;
+  private final Iterator<JsonElement> iterator;
+  private final JsonObject json;
+  private final Map<String, String> fieldsMapping;
+
+  JsonPage(BaseHttpSourceConfig config, String responseBody) {
+    json = JSONUtil.toJsonObject(responseBody);
+    JSONUtil.JsonQueryResponse queryResponse = JSONUtil.getJsonElementByPath(json, config.getResultPath());
+    insideElementJsonPathPart = queryResponse.getUnretrivedPath();
+    iterator = queryResponse.getAsJsonArray().iterator();
+    fieldsMapping = config.getFullFieldsMapping();
+  }
+
+  @Override
+  public boolean hasNext() {
+    return iterator.hasNext();
+  }
+
+  @Override
+  public String next() {
+    JsonObject currentJsonObject = iterator.next().getAsJsonObject();
+
+    JsonObject resultJson = new JsonObject();
+    for (Map.Entry<String, String> entry : fieldsMapping.entrySet()) {
+      String schemaFieldName = entry.getKey();
+      String fieldPath = insideElementJsonPathPart + "/" + StringUtils.stripStart(entry.getValue(), "/");
+
+      JSONUtil.JsonQueryResponse queryResponse = JSONUtil.getJsonElementByPath(currentJsonObject, fieldPath);
+      queryResponse.assertFullyRetrieved();
+
+      resultJson.add(schemaFieldName, queryResponse.get());
+    }
+    return resultJson.toString();
+  }
+
+  /**
+   * Get primitive from json by json path or return null if not found.
+   * If element found is not a primitive (object or array) exception is thrown.
+   *
+   * @param path a json path. E.g. "/city/schools/students"
+   * @return a primitive converted to string
+   */
+  @Override
+  public String getPrimitiveByPath(String path) {
```

This is a BasePage method? What does it mean for the delimited formats?

I also don't see where this is used, when is it called?

---

This is called by pagination iterators: "Link In Response Body" and "Link in Response Token". To retrieve next page url/token. 

Delimited and text types will fail with `Page format '%s' does not support searching by path`
However this exception will never actually happen.
Since if user chooses pagination via "Link In Response Body" or "Link in Response Token" and type like text,csv,tsv,blob config validation will fail early.